### PR TITLE
[4.0] httpheaders Implement the suggestion by @brianteeman

### PR DESCRIPTION
**Pull request by zero-24**
### Summary of Changes

Implement the list suggested by @brianteeman 

### Testing Instructions

Confirm that the custom headers are working

### Expected result

custom headers are working

### Actual result

custom headers are not working

### Note

@brianteeman please suggest the language changes. Maybe dropping the description completely?

And please tell me which strings from here: https://github.com/joomla/joomla-cms/pull/20459 do you want to keep or change. I'm happy to implement.

### CODE REVIEW

**Comment by zero-24** [comment]
Ah and @brianteeman do you agree that we don't allow the values to be translated or do you want me to move them to language strings?

**Comment by brianteeman** [comment]
Correct they don't need to be translated. Not even sure they could be

**Comment by brianteeman** [comment]
Not sure if its my branch but I am now missing some strings. Can you check please

We don't need the PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER_DESC anymore and all the strings for the options can go as well as things like X-XSS-Protection should not be translated.

Finally can you change the name to System HTTP Headers - we have a space bar on an english keybard :)

**Comment by zero-24** [comment]
> Correct they don't need to be translated. Not even sure they could be

Done also for the referrerpolicy rules.

> Not sure if its my branch but I am now missing some strings. Can you check please

What do you mean? did you pull my branch?

> We don't need the PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER_DESC anymore and all the strings for the options can go as well as things like X-XSS-Protection should not be translated.

XSS Protection? I guess you mean the referrerpolicy?

> Finally can you change the name to System HTTP Headers - we have a space bar on an english keybard :)

Done. :)

**Comment by brianteeman** [comment]
I fixed my branch and now this is all good - my concern with the language strings can be illustrated by this string
PLG_SYSTEM_HTTPHEADERS_REFERRERPOLICY="<a href='https://scotthelme.co.uk/a-new-security-header-referrer-policy/' target='_blank'>Referrer-Policy</a>"

I am concerned that "Referrer-Policy" would be translated by the TT and it absolutely must not be. The same applies to all the other headers which must not be translated in the text.

From our experience with the "installation" folder string the TT will ignore any notes we add to the language file not to translate these words and the only solution we found for that one was to remove "installation" from the language file strings and just have a %s 

Its not ideal but i cant think of any other way to prevent these strings from being translated

**Comment by zero-24** [comment]
>I am concerned that "Referrer-Policy" would be translated by the TT and it absolutely must not be. The same applies to all the other headers which must not be translated in the text.

Agree my point is that we use to include the link with more infos using that language string. Do you think we should drop them?

**Comment by brianteeman** [comment]
well its not just there it is also in the post installation message and some of the _DESC

other than the ugly %s i dont know what else to suggest

**Comment by zero-24** [comment]
> well its not just there it is also in the post installation message and some of the _DESC

hmm my problem is that i don't know how to append them using %s as the postinstall is hardcoded and the same apply to the xml labels and description.

**Comment by brianteeman** [comment]
maybe ask the TT Leader for some input?

**Comment by zero-24** [comment]
> maybe ask the TT Leader for some input?

@infograf768 can you please take a look here? Thanks!

**Comment by brianteeman** [comment]
I meant https://volunteers.joomla.org/joomlers/267-ilagnayeru-mig-manickam

**Comment by zero-24** [comment]
Ping @imanickam too.

**Comment by imanickam** [comment]
Good Morning!

Reviewed the xml files and I do agree that the values need not be derived from the language strings.  Whereas, the description text for the options could use language strings (e.g. JYES, JNO).

Brian's concern is very valid that it should not happen just like the word 'installation' has been translated in the installation ini files. However, I feel that this case is slightly different.
  
We can have the policy names (e.g., Referrer-Policy, X-XSS-Protection) as distinct language strings with language constants identifying the policy name so that they are grouped in one place in the language file, and have a comment not to translate the policy names.

Example:
; Please DO NOT TRANSLATE the language values for the language strings starting with PLG_SYSTEM_HTTPHEADERS_POLICY_NAME
PLG_SYSTEM_HTTPHEADERS_POLICY_NAME_REFERRERPOLICY="Referrer-Policy"
PLG_SYSTEM_HTTPHEADERS_POLICY_NAME_XXSSPROTECTION="X-XSS-Protection"

As these language strings would be used for display purpose only, IMHO, it should not affect any outcome.

As for as using placeholders within the language strings of Post Install messages, let me request JM to offer suggestions.

**Comment by infograf768** [comment]
Trying to find a solution.

In the mean while, please change EOL from Windows to Unix for the `en-GB.plg_system_httpheaders.ini` file

Also, when debug lang is on, I get these JGLOBAL strings constants for the tabs :
screen shot 2018-05-21 at 08 48 44

These JGLOBAL strings can replace the 
```
COM_PLUGINS_HSTS_FIELDSET_LABEL="Strict-Transport-Security (HSTS)"
COM_PLUGINS_CSP_FIELDSET_LABEL="Content-Security-Policy (CSP)"
```
so that we get the same result when debug lang is on or not.

**Comment by infograf768** [comment]
Some thoughts as they come:

This is a totally new plugin for a new version of Joomla where we will ask TTs to re-apply as TTs and where the packs will be quite different as 90% of the ini files will have to be largely modified. 

Therefore the issue can't be compared with the `installation` issue which was a > 10 years old one. What I mean is that the "new" TTs will be strongly briefed about what should and what should not be translated in these new strings.

NOTE:
For forms (not for postinstall), this shows we could improve the field translations.
What I am thinking of is something like this:
add new $element(s) (or many, incremented) `var-label="MYSTRING_VAR_LABEL"` and `var-description="MYSTRING_VAR_DESC"`
Add these new strings in the ini file concerned. Group them.
Then, when present in the field, translate them with a `JText_::_`
Then, still if present in the field, instead of using a `JText::_` for the label or description, use a `JText::sprintf` adding this new variable (or multi). 
Looks like we could do that in `FormField`

**Comment by brianteeman** [comment]
> This is a totally new plugin for a new version of Joomla where we will ask TTs to re-apply as TTs and where the packs will be quite different as 90% of the ini files will have to be largely modified.

I was not aware of that so maybe we can manage with just the "do not translate this" comment

**Comment by infograf768** [comment]
It would have to be more precise depending on the strings as the words not to translate may be part of stuff which need translating. 
Presenting these words not to translate between quotes helps Translators to understand that they do not need to be translated.

Example:
original string
`PLG_SYSTEM_HTTPHEADERS_XCONTENTTYPEOPTIONS="<a href='https://scotthelme.co.uk/hardening-your-http-response-headers/#x-content-type-options' target='_blank'>X-Content-Type-Options</a>"`
could be
`PLG_SYSTEM_HTTPHEADERS_XCONTENTTYPEOPTIONS="<a href='https://scotthelme.co.uk/hardening-your-http-response-headers/#x-content-type-options' target='_blank'>\"X-Content-Type-Options\"</a>"`

**Comment by brianteeman** [comment]
In a sentence then yes it does but where it is the entire label as it is here then it is not correct for the UI

**Comment by zero-24** [comment]
Thanks i have just commited the "do not translate" comments & used the JGLOBAL strings as suggested please review. https://github.com/joomla/joomla-cms/pull/20491/commits/38abbbdb13b8e57cc1e039adc53b774abc7d963c

**Comment by brianteeman** [comment]
@zero-24 its not that simple

"HTTP Header" can not be translated - nor can ANY of the technical terms anywhere

**Comment by zero-24** [comment]
> "HTTP Header" can not be translated - nor can ANY of the technical terms anywhere

I have just extended the comments. :)

File: administrator/language/en-GB/en-GB.plg_system_httpheaders.ini [hunk 1]
```diff
@@ -3,14 +3,17 @@
...
+JGLOBAL_FIELDSET_CSP="Content-Security-Policy (CSP)"
+PLG_SYSTEM_HTTPHEADERS="System - HTTP Headers"
+; Please do not translate the word HTTP Header in the following 3 language strings
 PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER="Define additional HTTP Header"
-PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER_DESC="Using this you can set also different values from the default ones and also additional headers when it is supported. The supported headers are: <br><ul><li>Strict-Transport-Security</li><li>Content-Security-Policy</li><li>Content-Security-Policy-Report-Only</li><li>X-Frame-Options</li><li>X-XSS-Protection</li><li>X-Content-Type-Options</li><li>Referrer-Policy</li><li>Expect-CT</li></ul>"
 PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER_KEY="HTTP Header"
 PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER_VALUE="HTTP Header Value"
+; Please do not translate the following language string
 PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY="<a href='https://scotthelme.co.uk/content-security-policy-an-introduction' target='_blank'>Content Security Policy (CSP)</a>"
+; Please do not translate the names of the http headers in the following 2 language strings
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_system_httpheaders.ini]: copy paste error should be "Content-Security-Policy" not http headers

File: administrator/language/en-GB/en-GB.plg_system_httpheaders.ini [hunk 2]
```diff
@@ -3,14 +3,17 @@
...
 PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_DESC="With this option all values for the 'Content-Security-Policy' can be set individually"
 PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_REPORT_ONLY_DESC="This option uses the header 'Content-Security-Policy-Report-Only' instead of 'Content-Security-Policy"
 PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_REPORT_ONLY="Report-Only"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_system_httpheaders.ini]: do not translate

File: administrator/language/en-GB/en-GB.plg_system_httpheaders.ini [hunk 3]
```diff
@@ -3,14 +3,17 @@
...
-PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER_DESC="Using this you can set also different values from the default ones and also additional headers when it is supported. The supported headers are: <br><ul><li>Strict-Transport-Security</li><li>Content-Security-Policy</li><li>Content-Security-Policy-Report-Only</li><li>X-Frame-Options</li><li>X-XSS-Protection</li><li>X-Content-Type-Options</li><li>Referrer-Policy</li><li>Expect-CT</li></ul>"
 PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER_KEY="HTTP Header"
 PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER_VALUE="HTTP Header Value"
+; Please do not translate the following language string
 PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY="<a href='https://scotthelme.co.uk/content-security-policy-an-introduction' target='_blank'>Content Security Policy (CSP)</a>"
+; Please do not translate the names of the http headers in the following 2 language strings
 PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_DESC="With this option all values for the 'Content-Security-Policy' can be set individually"
 PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_REPORT_ONLY_DESC="This option uses the header 'Content-Security-Policy-Report-Only' instead of 'Content-Security-Policy"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_system_httpheaders.ini]: do not translate Content-Security-Policy-Report-Only

File: administrator/language/en-GB/en-GB.plg_system_httpheaders.ini [hunk 4]
```diff
@@ -19,29 +22,27 @@ PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_VALUES_DIRECTIVE="Policy directive"
 PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_VALUES_VALUE="Value"
 PLG_SYSTEM_HTTPHEADERS_HEADER_CLIENT="Client"
 PLG_SYSTEM_HTTPHEADERS_HEADER_CLIENT_BOTH="Both"
+; Please do not translate the following language string
 PLG_SYSTEM_HTTPHEADERS_HSTS="<a href='https://hstspreload.org' target='_blank'>HTTP Strict Transport Security (HSTS)</a>"
+; Please do not translate the names of the http headers in the following language string
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_system_httpheaders.ini]: copy paste error should be "strict transport security" not http headers

File: administrator/language/en-GB/en-GB.plg_system_httpheaders.ini [hunk 5]
```diff
@@ -19,29 +22,27 @@ PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_VALUES_DIRECTIVE="Policy directive"
 PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_VALUES_VALUE="Value"
 PLG_SYSTEM_HTTPHEADERS_HEADER_CLIENT="Client"
 PLG_SYSTEM_HTTPHEADERS_HEADER_CLIENT_BOTH="Both"
+; Please do not translate the following language string
 PLG_SYSTEM_HTTPHEADERS_HSTS="<a href='https://hstspreload.org' target='_blank'>HTTP Strict Transport Security (HSTS)</a>"
+; Please do not translate the names of the http headers in the following language string
 PLG_SYSTEM_HTTPHEADERS_HSTS_DESC="This option activates 'Strict Transport Security' and allows the configuration of the values"
 PLG_SYSTEM_HTTPHEADERS_HSTS_MAXAGE="Maximum registration time"
 PLG_SYSTEM_HTTPHEADERS_HSTS_MAXAGE_DESC="This option sets the time for 'max-age', it is specified in seconds. The default value is 31536000, which corresponds to one year"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_system_httpheaders.ini]: do not translate max-age

File: administrator/language/en-GB/en-GB.plg_system_httpheaders.ini [hunk 6]
```diff
@@ -19,29 +22,27 @@ PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_VALUES_DIRECTIVE="Policy directive"
...
 PLG_SYSTEM_HTTPHEADERS_HSTS_MAXAGE="Maximum registration time"
 PLG_SYSTEM_HTTPHEADERS_HSTS_MAXAGE_DESC="This option sets the time for 'max-age', it is specified in seconds. The default value is 31536000, which corresponds to one year"
+; Please do not translate the following language string
 PLG_SYSTEM_HTTPHEADERS_HSTS_PRELOAD="Preload"
 PLG_SYSTEM_HTTPHEADERS_HSTS_PRELOAD_DESC="This option activates the opt-in for inclusion in so-called browser preload lists"
 PLG_SYSTEM_HTTPHEADERS_HSTS_PRELOAD_NOTE="Important"
 PLG_SYSTEM_HTTPHEADERS_HSTS_PRELOAD_NOTE_DESC="HSTS means that your domain can no longer be called without HTTPS. Once added to the preload list, this is not easy to undo. Domains can be removed, but it takes months for users to make a change with a browser update.<br><strong>This option is very important to prevent 'man-in-the-middle attacks', so it should be activated in any case, but only if you are sure that HTTPS is supported for domain and all subdomains in the long run! The value for'maximum recording time' must be set to 63072000 (2 years) for recording.</strong>"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_system_httpheaders.ini]: do not translate "maximum recording time"

File: administrator/language/en-GB/en-GB.plg_system_httpheaders.ini [hunk 7]
```diff
@@ -19,29 +22,27 @@ PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_VALUES_DIRECTIVE="Policy directive"
...
-PLG_SYSTEM_HTTPHEADERS_REFERRERPOLICY_NO_REFERRER_WHEN_DOWNGRADE="no-referrer-when-downgrade"
-PLG_SYSTEM_HTTPHEADERS_REFERRERPOLICY_SAME_ORIGIN="same-origin"
-PLG_SYSTEM_HTTPHEADERS_REFERRERPOLICY_ORIGIN="origin"
-PLG_SYSTEM_HTTPHEADERS_REFERRERPOLICY_STRICT_ORIGIN="strict-origin"
-PLG_SYSTEM_HTTPHEADERS_REFERRERPOLICY_ORIGIN_WHEN_CROSS_ORIGIN="origin-when-cross-origin"
-PLG_SYSTEM_HTTPHEADERS_REFERRERPOLICY_STRICT_ORIGIN_WEHN_CROSS_ORIGIN="strict-origin-when-cross-origin"
-PLG_SYSTEM_HTTPHEADERS_REFERRERPOLICY_UNSAFE_URL="unsafe-url"
 PLG_SYSTEM_HTTPHEADERS_XCONTENTTYPEOPTIONS="<a href='https://scotthelme.co.uk/hardening-your-http-response-headers/#x-content-type-options' target='_blank'>X-Content-Type-Options</a>"
 PLG_SYSTEM_HTTPHEADERS_XFRAMEOPTIONS="<a href='https://scotthelme.co.uk/hardening-your-http-response-headers/#x-frame-options' target='_blank'>X-Frame-Options</a>"
 PLG_SYSTEM_HTTPHEADERS_XML_DESCRIPTION="This Plugin helps you to set the HTTP Security Headers"
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.plg_system_httpheaders.ini]: do not translate "HTTP Security Headers"

**Review by brianteeman — CHANGES_REQUESTED** [review]

**Comment by zero-24** [comment]
Done thanks @brianteeman

**Comment by infograf768** [comment]
Is `Maximum recording time` same as `Maximum registration time` ?
if it is let's choose one of them only

File: administrator/language/en-GB/en-GB.plg_system_httpheaders.ini (hunk 6 above)
Comment (infograf768) [reply administrator/language/en-GB/en-GB.plg_system_httpheaders.ini]: Why not translate that?

**Comment by infograf768** [comment]
@brianteeman 
Why `maximum recording time` should not be translated?

File: administrator/language/en-GB/en-GB.plg_system_httpheaders.ini (hunk 5 above)
Comment (Quy) [reply administrator/language/en-GB/en-GB.plg_system_httpheaders.ini]: Let's remove `it is specified ` to be `This option sets the time for 'max-age' in seconds.` or `This option sets the time for 'max-age' specified in seconds.`

**Comment by brianteeman** [comment]
Sorry I might be confusing the maximum recording time - I thought that was the name of the actual attribute. Looking further if i understand the code correctly this refers to the attribute "max-age". That attribute should not be translated AND "maximum recording time" is not a correct description of "max-age"

the attribute "max-age" refers to how long the policy should remain in effect https://tools.ietf.org/html/rfc6797#section-6.1.1 so I am not sure why you have written it as "maximum recording time" 

it should just be the untranslatable "max-age"

**Comment by zero-24** [comment]
pushed thanks!

**Review by brianteeman — APPROVED** [review]

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 0aecc2f87e58b19ae786aabd08560f71985ec9d3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20491">issues.joomla.org/tracker/joomla-cms/20491</a>.</sub>

File: administrator/language/en-GB/en-GB.plg_system_httpheaders.ini [hunk 8]
```diff
@@ -3,45 +3,51 @@
...
 PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER_KEY="HTTP Header"
 PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER_VALUE="HTTP Header Value"
+; Please do not translate the following language string
 PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY="<a href='https://scotthelme.co.uk/content-security-policy-an-introduction' target='_blank'>Content Security Policy (CSP)</a>"
+; Please do not translate 'Content-Security-Policy' & 'Content-Security-Policy-Report-Only' in the following 2 language strings
 PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_DESC="With this option all values for the 'Content-Security-Policy' can be set individually"
 PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_REPORT_ONLY_DESC="This option uses the header 'Content-Security-Policy-Report-Only' instead of 'Content-Security-Policy"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.plg_system_httpheaders.ini]: Missing ending `'`

File: administrator/language/en-GB/en-GB.plg_system_httpheaders.ini [hunk 9]
```diff
@@ -3,45 +3,51 @@
...
+; Please do not translate the following language string
 PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY="<a href='https://scotthelme.co.uk/content-security-policy-an-introduction' target='_blank'>Content Security Policy (CSP)</a>"
+; Please do not translate 'Content-Security-Policy' & 'Content-Security-Policy-Report-Only' in the following 2 language strings
 PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_DESC="With this option all values for the 'Content-Security-Policy' can be set individually"
-PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_REPORT_ONLY_DESC="This option uses the header 'Content-Security-Policy-Report-Only' instead of 'Content-Security-Policy"
+PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_REPORT_ONLY_DESC="This option uses the header 'Content-Security-Policy-Report-Only' instead of 'Content-Security-Policy'"
+; Please do not translate the following language string
 PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_REPORT_ONLY="Report-Only"
 PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_VALUES="Add Value"
 PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_VALUES_DIRECTIVE="Policy directive"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.plg_system_httpheaders.ini]: Capitalize `d`

File: plugins/system/httpheaders/subform/customhttpheaders.xml [hunk 10]
```diff
@@ -1,10 +1,22 @@
...
 		label="PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER_KEY"
-	/>
+		default="no-referrer-when-downgrade"
+		validate="options"
+		>
+		<option value="strict-transport-security">Strict-Transport-Security</option>
+		<option value="content-security-policy">Content-Security-Policy</option>
+		<option value="content-security-policy-report-only">Content-Security-Policy-Report-Only</option>
+		<option value="x-frame-options">X-Frame-Options</option>
+		<option value="x-xss-protection">X-XSS-Protection</option>
```
Comment (Quy) [inline plugins/system/httpheaders/subform/customhttpheaders.xml:14]: https://scotthelme.co.uk/hardening-your-http-response-headers/#x-xss-protection has it as `X-Xss-Protection` (not XSS)

Comment (zero-24) [reply plugins/system/httpheaders/subform/customhttpheaders.xml:14]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-XSS-Protection

:) So X-XSS is fine :)

File: administrator/language/en-GB/en-GB.plg_system_httpheaders.ini (hunk 9 above)
Comment (zero-24) [reply administrator/language/en-GB/en-GB.plg_system_httpheaders.ini]: fixed

File: plugins/system/httpheaders/subform/customhttpheaders.xml [hunk 11]
```diff
@@ -1,10 +1,22 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <form>
+
 	<field
 		name="key"
-		type="text"
+		type="list"
 		label="PLG_SYSTEM_HTTPHEADERS_ADDITIONAL_HEADER_KEY"
-	/>
+		default="no-referrer-when-downgrade"
```
Comment (Quy) [inline plugins/system/httpheaders/subform/customhttpheaders.xml]: Not valid `no-referrer-when-downgrade`.

Comment (zero-24) [reply plugins/system/httpheaders/subform/customhttpheaders.xml]: fixed

File: administrator/language/en-GB/en-GB.plg_system_httpheaders.ini [hunk 12]
```diff
@@ -3,45 +3,51 @@
...
+; Please do not translate 'max-age' in the following language string
+PLG_SYSTEM_HTTPHEADERS_HSTS_PRELOAD_NOTE_DESC="HSTS means that your domain can no longer be called without HTTPS. Once added to the preload list, this is not easy to undo. Domains can be removed, but it takes months for users to make a change with a browser update.<br><strong>This option is very important to prevent 'man-in-the-middle attacks', so it should be activated in any case, but only if you are sure that HTTPS is supported for domain and all subdomains in the long run! The value for 'max-age' must be set to 63072000 (2 years) for recording.</strong>"
 PLG_SYSTEM_HTTPHEADERS_HSTS_SUBDOMAINS_DESC="HSTS should also be enabled <strong>for subdomains</strong> usually the subdomain 'www' is taken into account when creating the SSL certificate. If further subdomains are used, please note that they are also provided with a valid SSL certificate."
 PLG_SYSTEM_HTTPHEADERS_HSTS_SUBDOMAINS="Also for subdomains"
```
Comment (Quy) [inline administrator/language/en-GB/en-GB.plg_system_httpheaders.ini:38]: How about `Include Subdomains`?

**Comment by Quy** [comment]
I know this is beyond the scope of this PR, how about making `Site`, `Administrator`, `Both ` a dropdown instead radio buttons?

**Comment by infograf768** [comment]
@zero-24 
If this https://github.com/joomla/joomla-cms/pull/20525 is merged, we could simplify your strings by grouping the values not to translate in simple strings and then using a placeholder where fit.

**Comment by infograf768** [comment]
Even better. Let's take this label
PLG_SYSTEM_HTTPHEADERS_HSTS_MAXAGE="max-age"

we can simply do
PLG_SYSTEM_HTTPHEADERS_HSTS_MAXAGE="%s"
and add in the field xml
varLabel="max-age"
and TTs will never mistake.

**Comment by brianteeman** [comment]
That only helps where the term is the only thing in the string

**Comment by infograf768** [comment]
> That only helps where the term is the only thing in the string

Indeed. It is a bonus. But the feature let's use the variable from another string  as everybody understood... Example:

`PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY="<a href='https://scotthelme.co.uk/content-security-policy-an-introduction' target='_blank'>Content Security Policy (CSP)</a>"` 
can be replaced by
`PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY="<a href='https://scotthelme.co.uk/content-security-policy-an-introduction' target='_blank'>%s</a>"`

and then we use in the field
```
				<field
					name="contentsecuritypolicy"
					type="radio"
					label="PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY"
					description="PLG_SYSTEM_HTTPHEADERS_CONTENTSECURITYPOLICY_DESC"
					class="switcher"
					default="0"
					varLabel="JGLOBAL_FIELDSET_CSP"
					>
					<option value="0">JDISABLED</option>
					<option value="1">JENABLED</option>
				</field>
```
Where `JGLOBAL_FIELDSET_CSP` (already present string) has the value `"Content-Security-Policy (CSP)"`

Same evidently for the description and other strings in the ini files, and not only for this plugin.

**Comment by zero-24** [comment]
Until that PR is merged i think we should not block this PR and get some tests + merge? What do you think?

**Comment by infograf768** [comment]
As merging https://github.com/joomla/joomla-cms/pull/20525 would imply many changes in the ini strings, not sure we should.
If both people here in favor or immediate merging were just testing 20525, we could get further pretty fast. It is not as we would be in such a hurry that this PR here has to get in at once because of a urgent release.

**Comment by brianteeman** [comment]
I guess this will need updating when #20525 is merged (it is RTC)

**Comment by wilsonge** [comment]
Please remove 

```
JGLOBAL_FIELDSET_HSTS="Strict-Transport-Security (HSTS)"
JGLOBAL_FIELDSET_CSP="Content-Security-Policy (CSP)"
```

from the language files and just hardcode them in the XML - if they are standalone strings like this that should not be translated them don't put them in language files.

**Comment by zero-24** [comment]
Thanks implemented now @wilsonge 👍

**Comment by brianteeman** [comment]
CSP is not to be translated - source 
https://developer.mozilla.org/fr/docs/Web/HTTP/CSP

**Comment by zero-24** [comment]
> CSP is not to be translated - source
https://developer.mozilla.org/fr/docs/Web/HTTP/CSP

@brianteeman I have just looked at the current strings. ATM we only have one line with `CSP`
https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/language/en-GB/en-GB.plg_system_httpheaders.ini#L15
=> The complete string is marked as not to be translated.

The others are Hardcoded now in the XML like George suggested. I'm still missing something?

**Comment by infograf768** [comment]
@wilsonge 
It is pretty frustrating that https://github.com/joomla/joomla-cms/pull/20525 has been RTC for a while and not merged, which would have simplified this PR here and be quite useful in future other cases.
It would really help volunteers if a decision had been taken about it and not just ignoring it.

**Comment by zero-24** [comment]
When #20525 is merged we can do a PR that implements that.

**Comment by zero-24** [comment]
This has been merged so we can move forward with the csp integration.

**Pull request merged**